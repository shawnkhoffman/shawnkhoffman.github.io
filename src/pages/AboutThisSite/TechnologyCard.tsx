import React from 'react';

interface TechnologyCardProps {
  icon: React.ReactNode;
  title: string;
  description: string;
  link: string;
}

const TechnologyCard: React.FC<TechnologyCardProps> = ({ icon, title, description, link }) => {
  return (
    <a
      href={link}
      target="_blank"
      rel="noopener noreferrer"
      className="flex flex-col items-center text-center hover:bg-base-200 p-4 rounded-lg transition duration-300"
      aria-label={`Learn more about ${title}`}
    >
      <span aria-hidden="true">{icon}</span>
      <h3 className="font-semibold">{title}</h3>
      <p className="text-sm">{description}</p>
    </a>
  );
};

export default TechnologyCard;