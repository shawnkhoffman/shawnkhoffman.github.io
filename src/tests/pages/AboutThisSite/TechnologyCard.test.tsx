import { render, cleanup } from '@testing-library/react';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import TechnologyCard from '../../../pages/AboutThisSite/TechnologyCard';
import '@testing-library/jest-dom';

const MockIcon = () => <svg data-testid="mock-icon" />;

describe('TechnologyCard', () => {
  const defaultProps = {
    icon: <MockIcon />,
    title: 'React',
    description: 'A JavaScript library for building user interfaces.',
    link: 'https://reactjs.org/',
  };

  let container: HTMLElement;

  beforeEach(() => {
    cleanup();
    const result = render(<TechnologyCard {...defaultProps} />);
    container = result.container;
  });

  afterEach(() => {
    cleanup();
  });

  describe('Rendering', () => {
    it('renders all required elements', () => {
      expect(container.querySelector('[data-testid="mock-icon"]')).toBeInTheDocument();
      expect(container.querySelector('h3')).toHaveTextContent(defaultProps.title);
      expect(container.querySelector('p')).toHaveTextContent(defaultProps.description);
    });

    it('matches snapshot', () => {
      expect(container).toMatchSnapshot();
    });
  });

  describe('Structure', () => {
    it('maintains proper content structure', () => {
      const link = container.querySelector('a');
      expect(link).toBeInTheDocument();
      expect(link).toContainElement(container.querySelector('h3'));
      expect(link).toContainElement(container.querySelector('p'));
      expect(link).toContainElement(container.querySelector('[data-testid="mock-icon"]'));
    });
  });

  describe('Interactions', () => {
    it('opens link in new tab when clicked', () => {
      const link = container.querySelector('a');
      expect(link).toHaveAttribute('target', '_blank');
      expect(link).toHaveAttribute('href', defaultProps.link);
    });

    it('has security attributes for external link', () => {
      const link = container.querySelector('a');
      expect(link).toHaveAttribute('rel', 'noopener noreferrer');
    });
  });

  describe('Accessibility', () => {
    it('provides accessible name for the link', () => {
      const link = container.querySelector('a');
      expect(link).toHaveAccessibleName(`Learn more about ${defaultProps.title}`);
    });

    it('hides decorative icon from screen readers', () => {
      const iconWrapper = container.querySelector('[data-testid="mock-icon"]')?.closest('[aria-hidden="true"]');
      expect(iconWrapper).toHaveAttribute('aria-hidden', 'true');
    });

    it('is keyboard focusable', () => {
      const link = container.querySelector('a');
      if (link) {
        link.focus();
        expect(link).toHaveFocus();
      }
    });
  });
});
