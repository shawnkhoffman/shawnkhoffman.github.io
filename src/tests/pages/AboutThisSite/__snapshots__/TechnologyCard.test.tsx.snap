// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TechnologyCard > Interactions > has security attributes for external link 1`] = `
<div>
  <a
    aria-label="Learn more about React"
    class="flex flex-col items-center text-center hover:bg-base-200 p-4 rounded-lg transition duration-300"
    href="https://reactjs.org/"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span
      aria-hidden="true"
    >
      <svg
        data-testid="mock-icon"
      />
    </span>
    <h3
      class="font-semibold"
    >
      React
    </h3>
    <p
      class="text-sm"
    >
      A JavaScript library for building user interfaces.
    </p>
  </a>
</div>
`;
